pkg/reconciler/kubequota/kubequota_controller.go

```diff
@@ -0,0 +1,352 @@
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+// processNextWorkItem waits for the queue to have a key available and then processes it.
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	raw, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := raw.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+
+	return true
+}
+
+// process processes a single key from the queue.
+func (c *Controller) process(ctx context.Context, key string) error {
+	// e.g. root:org<separator>ws
+	parent, name := clusters.SplitClusterAwareKey(key)
+
+	// turn it into root:org:ws
+	clusterName := parent.Join(name)
+
+	_, err := c.getClusterWorkspace(key)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(2).InfoS("ClusterWorkspace not found - stopping quota controller for it", "clusterName", clusterName)
+
+			c.lock.Lock()
+			cancel, ok := c.cancelFuncs[clusterName]
+			if ok {
+				cancel()
+				delete(c.cancelFuncs, clusterName)
+			}
+			c.lock.Unlock()
+
+			return nil
+		}
+
+		return err
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, found := c.cancelFuncs[clusterName]
+	if found {
+		klog.V(4).InfoS("Quota controller already exists", "clusterName", clusterName)
+		return nil
+	}
+
+	klog.V(2).InfoS("Starting quota controller", "clusterName", clusterName)
+
+	ctx, cancel := context.WithCancel(ctx)
+	c.cancelFuncs[clusterName] = cancel
+
+	if err := c.startQuotaForClusterWorkspace(ctx, clusterName); err != nil {
+		return fmt.Errorf("error starting quota controller for cluster %q: %w", clusterName, err)
+	}
+
+	return nil
+}
+
+func (c *Controller) startQuotaForClusterWorkspace(ctx context.Context, clusterName logicalcluster.Name) error {
+	resourceQuotaControllerClient := c.kubeClusterClient.Cluster(clusterName)
+	resourceQuotaControllerDiscoveryClient := resourceQuotaControllerClient.Discovery()
+
+	discoveryFunc := resourceQuotaControllerDiscoveryClient.ServerPreferredNamespacedResources
+
+	dynamicClient := c.dynamicClusterClient.Cluster(clusterName)
+
+	// TODO(ncdc): pass in a scoped factory when available, rather than creating a new factory per logical cluster.
+	// TODO(ncdc): could this be a partial metadata factory for everything except the few types that require full
+	// object details?
+	dynamicInformerFactory := dynamicinformer.NewFilteredDynamicSharedInformerFactoryWithOptions(
+		dynamicClient,
+		"",
+		nil,
+		cache.WithResyncPeriod(c.fullResyncPeriod),
+		cache.WithIndexers(cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}),
+	)
+	dynamicInformerFactory.Start(ctx.Done())
```

Comment:
this seriously worries me, IMO a nogo for the PR. At the very least push the plumbing outside into main server.go. Nested informer creation is bad.

Reply:
Agreed, this was to get it working.